packages/gensx-core/src/checkpoint.ts

```diff
@@ -1,61 +1,36 @@
 import { Buffer } from "node:buffer";
+import { createHash } from "node:crypto";
 import { join } from "node:path";
 import { promisify } from "node:util";
 import { gzip } from "node:zlib";
 
-import { STREAMING_PLACEHOLDER } from "./component.js";
-import { ExecutionContext } from "./context.js";
-import { ComponentOpts } from "./types.js";
+import {
+  CheckpointWriter,
+  ExecutionNode,
+  STREAMING_PLACEHOLDER,
+} from "./checkpoint-types.js";
 import { readConfig } from "./utils/config.js";
 import { USER_AGENT } from "./utils/user-agent.js";
 
-const gzipAsync = promisify(gzip);
-
-// Cross-platform UUID generation
-function generateUUID(): string {
-  try {
-    // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/no-unsafe-member-access
-    const crypto = (globalThis as any).crypto as { randomUUID: () => string };
-    return crypto.randomUUID();
-  } catch {
-    // Simple fallback for environments without crypto
-    return "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(/[xy]/g, (c) => {
-      const r = (Math.random() * 16) | 0;
-      const v = c === "x" ? r : (r & 0x3) | 0x8;
-      return v.toString(16);
-    });
-  }
-}
+export type { CheckpointWriter, ExecutionNode };
+export { STREAMING_PLACEHOLDER };
 
-export interface ExecutionNode {
-  id: string;
-  componentName: string;
-  parentId?: string;
-  startTime: number;
-  endTime?: number;
-  props: Record<string, unknown>;
-  output?: unknown;
-  children: ExecutionNode[];
-  metadata?: {
-    logs?: string[];
-    tokenCounts?: {
-      input: number;
-      output: number;
-    };
-    [key: string]: unknown;
-  };
-  componentOpts?: ComponentOpts;
-}
+const gzipAsync = promisify(gzip);
 
-export interface CheckpointWriter {
-  root?: ExecutionNode;
-  addNode: (node: Partial<ExecutionNode>, parentId?: string) => string;
-  completeNode: (id: string, output: unknown) => void;
-  addMetadata: (id: string, metadata: Record<string, unknown>) => void;
-  updateNode: (id: string, updates: Partial<ExecutionNode>) => void;
-  write: () => void;
-  waitForPendingUpdates: () => Promise<void>;
-  checkpointsEnabled: boolean;
+// Deterministic ID generation for checkpoint replay
+export function generateDeterministicId(
+  name: string,
+  props: Record<string, unknown>,
+  parentId?: string,
+): string {
+  // Simple but effective serialization for MVP
+  const propsStr = JSON.stringify(props, Object.keys(props).sort());
+  const propsHash = createHash("sha256")
+    .update(propsStr)
+    .digest("hex")
+    .slice(0, 16);
+
+  return `${parentId ?? "root"}:${name}:${propsHash}`;
```

Comment:
One thing this does not take into account (and I did not solve in my prototype either) is multiple child components with the props. We basically make the assumption that they are all the identical in that case.

I think we could solve this via a sequence counter, that increase for each component, either at the parent or workflow level, and include that in the hash (or it is the whole hash?).

Reply:
Those could work. A sequence number would be make it no longer a concern of the user. Javascript iteration is deterministic, so it should be reliable unless it is non-deterministic code.

Reply:
Are you sure we can guarantee sequence in the case where components are created in a loop and awaited via promise.all?